text/0000-gen-fn.md

```diff
@@ -0,0 +1,785 @@
+This will probably be fixed by special casing the generator impl if `Generator::Return = ()`, as we can trivially
+produce infinite values of the unit type.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+## New keyword
+
+In the 2024 edition we reserve `gen` as a keyword. Previous editions will use `k#gen` to get the same features.
+
+## Error handling
+
+`foo?` in `gen` blocks will stop iteration after the first error by desugaring to:
+
+```rust
+match foo.branch() {
+    ControlFlow::Break(err) => {
+        yield R::from_residual(err);
+        return;
+    },
+    ControlFlow::Continue(val) => val,
+}
+```
+
+This is the same behaviour that `collect::<Result<_, _>>()` performs
+on iterators over `Result`s.
+
+## Implementation
+
+This feature is mostly implemented via existing generators.
+We'll need additional desugarings and lots of work to get good diagnostics.
+
+### `gen fn`
+
+`gen fn` desugars to the function itself with the return type replaced by `impl Iterator<Item = $ret>` and its body wrapped in a `gen` block.
+A `gen fn`'s "return type" is its iterator's `yield` type.
+
+A `gen fn` captures all lifetimes and generic parameters into the `impl Iterator` return type (just like `async fn`).
+If more control over captures is needed, type alias impl trait can be used when it is stabilized.
+
+Like other uses of `impl Trait`, auto traits are revealed without being specified.
+
+### `gen` blocks
+
+`gen` blocks are the same as an unstable generator...
+
+* ...without arguments,
+* ...with an additional check forbidding holding borrows across `yield` points,
+* ...and with an automatic `Iterator` implementation.
+
+We'll probably be able to modularize the generator implementation and make it more robust on the implementation and diagnostics side for the `gen` block case, but I believe the initial implementation should be a HIR lowering to a generator and wrapping that generator in [`from_generator`][].
+
+## Fusing
+
+Special case the generator implementation if `Generator::Return = ()` to not panic, but
+repeatedly produce values of the unit type.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It's another language feature for something that can already be written entirely in user code.
+
+In contrast to `Generator`, `gen` blocks that produce `Iterator`s cannot hold references across `yield` points.
+See [`from_generator`][] which has an `Unpin` bound on the generator it takes to produce an `Iterator`.
+
+[`from_generator`]: https://doc.rust-lang.org/std/iter/fn.from_generator.html
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+## Keyword
+
+We could use `iter` as the keyword.
+I prefer `iter` because I connect generators with a more powerful scheme than plain `Iterator`s.
+The `Generator` trait can do everything that `iter` blocks and `async` blocks can do and more.
```

Comment:
This probably needs updating now that that trait is getting renamed to `Coroutine`.

Also see [Wikipedia](https://en.wikipedia.org/wiki/Generator_(computer_programming)) -- it seems that "generator" is indeed usually tied to iterators.

Reply:
The rename to `Coroutine` is now reflected throughout the document.